src/protocol/command.rs

```diff
@@ -1,10 +1,95 @@
 use super::argument_parser::Parser;
 use super::executable_path;
 use crate::R;
+use regex::Regex;
 use std::ffi::OsString;
 use std::path::PathBuf;
 use std::str;
 
+#[derive(Debug, Clone)]
+pub enum CommandMatcher {
+    Exact(Command),
+    RegexMatch(Regex),
```

Comment:
I wonder if it'd be better for multiple reasons to have `RegexMatch(String)` instead.

- Then we could definitely derive `PartialEq` and `Eq`.
- Also, we could then consider getting rid of the constructor functions below and just use `Exact(Command::new(...))` and `RegexMatch(...)`. And make those public as needed.

I think making enums part of an interface us often a good idea, because it gives developers a quick way to gain a lot of understanding about how something works.

Reply:
this makes sense to me, but there is a snag, maybe you have some advice for: 

`Regex::new` returns a `Result` which will error if there is an issue with the regex. IMO this error should occur at YAML parse time (i'll add a test for this). the options i can think of are:
* defer regex errors to run time (so the protocol test will always fail)
* run `Regex::new` at construction, bubble up the error, ignore the `Ok` result, reconstruct it at match time
* run `Regex::new` at construction, bubble up the error, save the `Ok` result
  * in the `enum` like `RegexMatch(String, Regex)` (but this doesn't fix the `PartialEq` issue)
  * in the `enum` like `RegexMatch(String, Box<Regex>)` (guessing here, not sure if this fixes the `PartialEq` issue)
  * using `lazy_static` as recommended here: https://github.com/rust-lang/regex#usage-avoid-compiling-the-same-regex-in-a-loop , but i have to doublecheck on how that actually works, and if it would work here

Reply:
I see. :| I think the error should happen at parse time, so I don't think we should defer that.

The second idea (calling `Regex::new` twice) is a bit wonky.

Using `Box<Regex>` doesn't solve the issue, see [here](https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-PartialEq%3CBox%3CT%3E%3E).

Using `lazy_static` feels even wonkier than calling it twice, I think.

What about this: Create a thin wrapper around `Regex` that has the original string and the compiled `Regex`. With `YourCoolWrapper::new() -> R<YourCoolWrapper>`. And then for that we implement `PartialEq` and `Eq` manually. But then for `CommandMatcher` we can derive it. Maybe it can be called `AnchoredRegex`, since that's the functionality that it'd add.

Reply:
i like that, will do